src/libpmemset/part.c

```diff
@@ -202,9 +224,31 @@ void
 pmemset_part_map_delete(struct pmemset_part_map **part_map)
 {
 	struct pmemset_part_map *pmap = *part_map;
-	pmem2_map_delete(&pmap->pmem2_map);
-	if (pmap->pmem2_reserv)
-		pmem2_vm_reservation_delete(&pmap->pmem2_reserv);
-	Free(pmap);
+	struct pmem2_vm_reservation *pmem2_reserv = pmap->pmem2_reserv;
+
+	void *rsv_addr = pmem2_vm_reservation_get_address(pmem2_reserv);
+	size_t rsv_size = pmem2_vm_reservation_get_size(pmem2_reserv);
+
+	struct pmem2_map *map;
+	/* find mapping lying at the end of the reservation */
+	pmem2_vm_reservation_map_find(pmem2_reserv, rsv_size - 1, 1, &map);
+	ASSERTne(map, NULL);
+
+	void *pmem2_map_addr = pmem2_map_get_address(map);
+	size_t pmem2_map_size = pmem2_map_get_size(map);
+
+	pmem2_map_delete(&map);
+
+	if (rsv_addr == pmem2_map_addr && rsv_size == pmem2_map_size) {
+		/* mapping covers whole vm reservation */
+		pmem2_vm_reservation_delete(&pmem2_reserv);
+		Free(pmap);
+	} else {
+		/* mapping covers part of the vm reservation */
```

Comment:
when does this happen? isn't the point of coalescing to have only one pmemset_part_map per reservation?